packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -120,41 +120,42 @@ contract Bridge is IBridge, ReentrancyGuard, Claimable
 
     modifier onlyFromExchangeOwner()
     {
-        require(msg.sender == exchangeOwner, "UNAUTHORIZED");
+        require(msg.sender == exchange.owner(), "UNAUTHORIZED");
```

Comment:
The reason to cache this is that this is/was x10 more expensive than a locally cached value.
However, now that we're after the Berlin HF and the gas prices have changed for calls/storage reads this may not be the case anymore. Because the gas cost new depends on other operations as well I don't know what the cheapest method is. I don't know yet if truffle supports Berlin gas costs, but if so I will measure both versions to see. I'll do the same with the token caching mechanism, that one may also be impacted in the same way.

Reply:
If you find out the cost is higher, I will revert the change to have a local owner (whose value will be set to the exchange owner).

Reply:
Brecht, how about we simply use  Claimable and allow the owenr to access onReceiveTransactions and trustConnector?

Reply:
Same thing with Converter, how about we also simply check the owner(), not the exchange's owner. I believe converters and bridge needs to be added as Agent, if we do that, we already authorized them.

Reply:
`onReceiveTransactions` needs to be called by the exchange owner contract otherwise it cannot be called together with a block. `trustConnector` could indeed be called by anyone.

For the Converter we use the exchange owner as the intermediate holder of the vault tokens, so when the callback is called from the exchange owner contract it is simply `msg.sender` like usual so no special logic is needed. This could work differently, but this still seems to be an easy and efficient way to do things.

> I believe converters and bridge needs to be added as Agent, if we do that, we already authorized them.

Only the Bridge is an Agent (because L2 transactions are approved in it). Converts are not agents because all user transactions are all just simple L2 trades signed by them on L2, so nothing special about them on L1.

Reply:
So it seems support for Berlin is sill missing in the tools we use... Still an open issue on the github of truffle/ganache-cli and also remix doesn't expose it yet. But support should be coming pretty soon.

My approximation for the cost is 2,500 gas when not caching it, because we are on the cheap path with all contracts/storage already needing to be accessed for the block submission anyway. The cached gas cost would be 2,100 gas, so a bit cheaper. The other option is use an immutable value for it, but that makes deploying the contract harder and the implementation would need to be updated for each exchange owner change, so doesn't seem like a good idea. Because of the very small difference it does seem like caching the value isn't worth it any more so keeping your change.

For the token caching, it may be cheaper not to used the cached versions for the deposits, but still use the for the connector calls. Easy to change so would leave it with caching for now, will update if necessary after I measure it.